### PR TITLE
feat: use fields to store xblock state

### PR DESCRIPTION
**Pull request by Ian2012**
## Description

This PR updates the implementation of the xblock to use xblock fields to store the state of the static mindmap and the student mindmap.

This PR fixes a bug in which if we add two instances of the xblock to the same subsection it would crash.

This PR does a major refactor simplifying code to get the context and to render the views

## Test instructions:

1. In the CMS
2. Create a static mindmap.
3. You should be able to edit it on the Edit page, which should be saved once you click Save.
4. Create an editable mindmap.
5. You should be able to edit it on the Edit page; students can only see it in the LMS.
6. When learners load the XBlock, they will get the initial Xblock configured in Studio and can start working on it and saving it.
7. Students should be able to see the instructions to use the xblock.
8. Students can reload the page and see their updated version of the xblock

# TODO

- Update docstrings
- Add tests

**Merge checklist:**
Check off if complete *or* not applicable:
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Fixup commits are squashed away
- [ ] Unit tests added/updated
- [ ] Manual testing instructions provided
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

### CODE REVIEW

**Comment by mariajgrimaldi** [comment]
Why is this a better approach to storing the students' mindmap than the previous one we already tested?

**Comment by Ian2012** [comment]
Advantages of this approach:

- The code is more maintainable as this is the recommended way to store xblock-related information
- We don't need to worry about handling storage-specific code (like having to remove the old file to update it).
- The implementation is simpler
- We don't depend on an external service (if configured)
- Serialization is faster as we use the database to store state and don't have to serialize a file and then convert it to JSON.
- It allows to duplicate, move and reorder the xblock and export the course with the xblock configured.

File: mindmap/static/html/mindmap.html [hunk 1]
```diff
@@ -1,24 +1,53 @@
...
-            {% if not is_static %}
-                <button class="save-button-student">{% trans "Save" %}</button>
-            {% endif %}
-            <div id="jsmind_container"></div>
-        </div>
-    {% endif %}
+
+<h1>{{display_name}}</h1>
+{% if editable and in_student_view %}
+    <p>Instrucctions</p>
```
Comment (mariajgrimaldi) [inline mindmap/static/html/mindmap.html:6]: Please, remove this and move it to another PR.

File: mindmap/static/js/src/mindmap.js [hunk 2]
```diff
@@ -64,13 +59,13 @@ function loadJSMind(callback) {
...
         // Load jsMind dynamically using $.getScript
-        $.getScript("https://unpkg.com/jsmind@latest/es6/jsmind.js")
+        $.getScript("https://unpkg.com/jsmind@0.6.4/es6/jsmind.js")
             .done(function () {
                 // Assign jsMind to the window object after it's loaded
                 window.jsMind = jsMind;
 
                 // Load jsMind.draggable-node dynamically using $.getScript
-                $.getScript("https://unpkg.com/jsmind@latest/es6/jsmind.draggable-node.js")
+                $.getScript("https://unpkg.com/jsmind@0.6.4/es6/jsmind.draggable-node.js")
```
Comment (mariajgrimaldi) [inline mindmap/static/js/src/mindmap.js:68]: why changing this?

File: mindmap/static/js/src/mindmapEdit.js [hunk 3]
```diff
@@ -1,22 +1,87 @@
 /* Javascript for MindMapXBlock. */
 
 // TODO: add notifications
-function MindMapXBlock(runtime, element) {
+function MindMapXBlock(runtime, element, context) {
```
Comment (mariajgrimaldi) [inline mindmap/static/js/src/mindmapEdit.js:4]: why changing this too?

**Comment by mariajgrimaldi** [comment]
I agree with using this approach with the static mindmap since we should maintain the block state. But the students' submission is not xblock-related information, so we managed it as a file submission as in edx-sga or edx-ora2. Do we know what's the max size for the dict field?

**Comment by mariajgrimaldi** [comment]
Please, don't add changes not related to the PR description. It gets a bit difficult to read.

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]
I added a few changes and questions. I like that we can use the mindmap JSON format to save it directly to the database, although I'm curious about the max size of the dict fields.

**Comment by mariajgrimaldi** [comment]
@BryanttV can you help me testing this out? Try using the same test cases you used for the previous PRs so we know this is working correctly. Thanks!

File: mindmap/static/js/src/mindmap.js (hunk 2 above)
Comment (Ian2012) [reply mindmap/static/js/src/mindmap.js:68]: It's not a good practice to use the latest release, I'm just pinning the latest compatible version

File: mindmap/static/js/src/mindmapEdit.js (hunk 3 above)
Comment (Ian2012) [reply mindmap/static/js/src/mindmapEdit.js:4]: I need the context similar to what `mindmap.js` does, it's to being able to render the mindmap in the edition panel

**Comment by mariajgrimaldi** [comment]
@BryanttV regarding your previous comment ^: I think we should always show the students mindmap unless is a static mindmap. What do you think? @Ian2012

File: mindmap/mindmap.py [hunk 4]
```diff
@@ -136,42 +178,13 @@ def student_view(self, _context=None) -> Fragment:
...
-        if statici18n_js_url:
-            frag.add_javascript_url(
-                self.runtime.local_resource_url(self, statici18n_js_url)
-            )
+        frag = self.load_fragment("mindmap", context)
 
         frag.add_javascript(self.resource_string("static/js/src/requiredModules.js"))
-        frag.add_javascript(self.resource_string("static/js/src/mindmap.js"))
-        frag.initialize_js('MindMapXBlock', json_args=js_context)
+        frag.initialize_js('MindMapXBlock', json_args=self.get_js_context(self.get_current_user(), context))
```
Comment (BryanttV) [inline mindmap/mindmap.py:186]: Couldn't the user be obtained from the `user` variable defined above?

**Comment by mariajgrimaldi** [comment]
@Ian2012: do we have an answer for [this](https://github.com/eduNEXT/xblock-mindmap/pull/10#pullrequestreview-1574049632) question?

**Review by mariajgrimaldi — APPROVED** [review]
LGTM. Before merging:
- Make tests pass
- Update the readme according to the new changes
- Update CHANGELOG/__init__ files
Then create a new release. Thanks!

**Review by mariajgrimaldi — APPROVED** [review]
Thank you both!

File: README.rst [hunk 5]
```diff
@@ -70,12 +70,17 @@ You can enable the Mind Map XBlock in the studio through the advanced settings.
...
 
 Fields
 ======
 - **Display name (String)**: Name of the component.
+- **Is a static mindmap? (Boolean)**: If this option is enabled, the student will not be able to edit the mind map.
+- **Mindmap (String)**: A JSON string that represents the mind map. Instructors will use an editor to
+  create the mind map which will be converted to a JSON string and saved in this field.
 
+In the case an instructor define a mindmap and the student is allowed to edit it, the student will see the initial Mindmap
+as the defined by the instructor and will be able to edit it.
```
[the remark this answers is not in the document]
Comment (Ian2012) [reply README.rst]: done

File: mindmap/static/js/src/mindmap.js [hunk 6]
```diff
@@ -64,13 +59,13 @@ function loadJSMind(callback) {
         callback();
     } else {
         // Load jsMind dynamically using $.getScript
-        $.getScript("https://unpkg.com/jsmind@latest/es6/jsmind.js")
+        $.getScript("https://unpkg.com/jsmind@0.6.4/es6/jsmind.js")
```
Comment (johnvente) [inline mindmap/static/js/src/mindmap.js:62]: Why are you changing `latest`? if the library is updated we won't have the latest version

Comment (Ian2012) [reply mindmap/static/js/src/mindmap.js:62]: Because the `latest` release can be a breaking change a break all the platforms with that version. Also, it's a good practice to pin the version of our dependencies

Comment (Ian2012) [reply mindmap/static/js/src/mindmap.js:62]: Sure, that can be included here https://github.com/eduNEXT/xblock-mindmap/pull/12
